File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.*;
import java.util.*;

public class Main {

    private static int[][] map;
    private static boolean[][] visited;
    private static final int[] dx = {1, -1, 0, 0};
    private static final int[] dy = {0, 0, 1, -1};
    private static int M;
    private static int N;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        // 테스트 케이스 갯수
        int T = Integer.parseInt(br.readLine());

        for(int i = 0; i < T; i++) {

            st = new StringTokenizer(br.readLine());

            // 필요한 지렁이 마리 수
            int cnt = 0;

            // 배추밭 가로길이
            M = Integer.parseInt(st.nextToken());
            // 배추밭 세로길이
            N = Integer.parseInt(st.nextToken());
            // 배추 위치 갯수
            int K = Integer.parseInt(st.nextToken());

            map = new int[M][N];
            visited = new boolean[M][N];

            // 배추들의 위치
            for(int j = 0; j < K; j++) {

                st = new StringTokenizer(br.readLine());

                // 배추 x, y 좌표값
                int xCoord = Integer.parseInt(st.nextToken());
                int yCoord = Integer.parseInt(st.nextToken());

                map[xCoord][yCoord] = 1;
            }

            for(int x = 0; x < M; x++) {
                for(int y = 0; y < N; y++) {

                    // 방문하지 않았고 배추가 심어져있을 경우에만 dfs
                    if(!visited[x][y] && map[x][y] == 1) {
                        cnt++;
                        dfs(x, y);
                    }
                }
            }
            sb.append(cnt).append("\n");
        }
        br.close();

        bw.write(sb.toString());
        bw.flush();
        bw.close();
    }

    private static void dfs(int x, int y) {

        // 지렁이 마리 수++ 하고 방문처리
        visited[x][y] = true;

        // 인접한 배추 확인
        for(int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            // 해당 좌표가 범위를 벗어나지 않고 방문하지 않았고 배추가 존재하는 경우일때 재귀
            if(nx > -1 && nx < M && ny > -1 && ny < N
                    && !visited[nx][ny] && map[nx][ny] == 1) {
                dfs(nx, ny);
            }
        }
    }
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.kt
import java.util.*
import java.io.*

val dx = arrayOf(1, -1, 0, 0)
val dy = arrayOf(0, 0, 1, -1)

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val T = readLine().toInt()
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    repeat(T) {
        var st = StringTokenizer(readLine())
        val M = st.nextToken().toInt() // 가로 길이
        val N = st.nextToken().toInt() // 세로 길이
        val K = st.nextToken().toInt() // 배추 갯수
        var warmCnt = 0

        val field = Array(N) {IntArray(M)} // 배추밭

        repeat(K) {
            st = StringTokenizer(readLine())
            val c = st.nextToken().toInt()
            val r = st.nextToken().toInt()
            field[r][c] = 1
        }

        for(r in 0 until N) {
            for(c in 0 until M) {
                if(field[r][c] == 1) {
                    dfs(r, c, N, M, field)
                    warmCnt++
                }
            }
        }
        bw.write("$warmCnt\n")
    }
    close()
    bw.flush()
    bw.close()
}
fun dfs(
    row: Int,
    column: Int,
    N: Int,
    M: Int,
    field : Array<IntArray>
) {
    for(i in 0 until 4) {
        val nr = row + dx[i]
        val nc = column + dy[i]

        // 배추밭 범위를 벗어나거나, 이미 방문했거나, 배추가 심어저 있지 않은 경우 continue
        if(nr < 0 || nc < 0 || nr > N - 1 || nc > M - 1 || field[nr][nc] != 1) {
            continue
        }
        field[nr][nc] = 0
        dfs(nr, nc, N, M, field)
    }
}